### ISSUE TITLE
A promise/BatteryManager *per browsing context*!?

### ISSUE THREAD

**Issue by domenic** [w3c/battery#2]
Is it really intentional that the same promise and battery manager are created per _browsing context_, instead of, say, per Document, or per Window?

This implies that the following will hold:

``` html
<!DOCTYPE html>

<iframe src="about:blank"></iframe>
<script>
"use strict";
const iframe = document.querySelector("iframe");

const originalNavigator = iframe.contentWindow.navigator;
const originalPromise = navigator.getBattery();

promise.then(originalManager => {
  iframe.onload = () => {
    // As expected, navigator changes:
    assert(iframe.contentWindow.navigator !== originalNavigator);

    // But the battery status promise does not!?!
    assert(iframe.contentWindow.navigator.getBattery() === originalPromise);

    // worse:
    iframe.contentWindow.navigator.getBattery().then(manager => {
      // the manager stays the same!!
      assert(manager === originalManager);

      // which means its prototype chain is broken!?
      assert(manager.__proto__ !== iframe.contentWindow.BatteryManager.prototype);
      assert(!(manager instanceof iframe.contentWindow.BatteryManager));
    });
  };
  iframe.src = "http://example.com";
});

// in fact, it's impossible to every have something be `instanceof BatteryManager`, even
// without iframe shenanigans, unless you are executing code in the initial about:blank document:
if (document.URL !== "about:blank" || document.history.length > 0) {
  navigator.getBattery().then(manager => {
    assert(!(manager instanceof BatteryManager));
  });
}
```

This seems very unlikely to be correct.

I notice this spec is going for PR. Is this behavior interoperable across multiple implementations?

**Comment by anssiko** [comment]
@domenic Thanks for the review and detailed bug report. The historical pointer where this feature was conceived is at: https://lists.w3.org/Archives/Public/public-device-apis/2014Jul/0003.html

I took your test case, fixed a typo (`s/promise.then/originalPromise.then/`), adapted `iframe.src`, instrumented and ran it. Results below.

Chrome 49:

```
assert(!(manager instanceof BatteryManager));
false
assert(iframe.contentWindow.navigator !== originalNavigator);
true
assert(iframe.contentWindow.navigator.getBattery() === originalPromise);
false
assert(manager === originalManager);
false
assert(manager.__proto__ !== iframe.contentWindow.BatteryManager.prototype);
false
assert(!(manager instanceof iframe.contentWindow.BatteryManager));
false
```

Firefox 45:

```
assert(!(manager instanceof BatteryManager));
false
assert(iframe.contentWindow.navigator !== originalNavigator);
false
assert(iframe.contentWindow.navigator.getBattery() === originalPromise);
false
assert(manager === originalManager);
false
assert(manager.__proto__ !== iframe.contentWindow.BatteryManager.prototype);
false
assert(!(manager instanceof iframe.contentWindow.BatteryManager));
false
```

It seems we probably want to `s/For each browsing context/For each Document/` to align the spec with the implementations as suggested. Quoting the [HTML Standard](https://html.spec.whatwg.org/multipage/browsers.html#windows) re the `Window` and `Document` object mapping:

> In general, there is a 1-to-1 mapping from the Window object to the Document object. There are two exceptions. First, a Window can be reused for the presentation of a second Document in the same browsing context, such that the mapping is then 1-to-2. This occurs when a browsing context is navigated from the initial about:blank Document to another, with replacement enabled. Second, a Document can end up being reused for several Window objects when the document.open() method is used, such that the mapping is then many-to-1.

Ping @mounirlamouri for comments.

@zqzhang It seems the w-p-t test suite does not have a test case for this. It'd be good to add this test case after we've settled on the issue.

**Comment by mounirlamouri** [comment]
Blink implements the feature with one promise per Document.

**Comment by anssiko** [comment]
See the proposed fix #3.

**Comment by zqzhang** [comment]
Good, I will make a test next week.
